core/app/models/spree/promotion_handler/promotion_batch_duplicator.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  module PromotionHandler
+    class PromotionBatchDuplicator < Spree::PromotionDuplicatorCore
+      def initialize(promotion, promotion_batch_id, random_string: nil, code: nil)
+        @promotion = promotion
+        @promotion_batch_id = promotion_batch_id
+        @random_string = random_string || generate_random_string(4)
+        @code = code
+      end
+
+      def duplicate
+        @new_promotion = @promotion.dup
+        @new_promotion.usage_limit = 1
+        @new_promotion.promotion_batch_id = @promotion_batch_id
+        @new_promotion.path = "#{@promotion.path}_#{@random_string}"
+        code_assignment
+        @new_promotion.stores = @promotion.stores
+
+        ActiveRecord::Base.transaction do
+          @new_promotion.save
+          copy_rules
+          copy_actions
+        end
+
+        @new_promotion
+      end
+
+      private
+
+      def code_assignment
+        return @new_promotion.generate_code=(true) unless @code
+
+        @new_promotion.code = @code
+      end
```

Comment:
Apart from the fact that this method is a bit redundant, I think it could be structured in a way that's much clearer:
```suggestion
      def code_assignment
        if @code
          @new_promotion.code = @code
        else
          @new_promotion.generate_code=(true)
        end
      end
```

This is subjective, but the `return`, the `generate_code=` call etc. make it harder to understand, it gets too "horizontal".

The version I sent is less clever, but IMO it's clearer what it's doing. Again - very subjective.